Use existing UTF-16 conversion logic in Java test generation, which was previously missing escapes

antlechner opened the pull request:
newline and carriage return need to be printed as \n and \r in Java code, respectively (as opposed to \u000a and \u000d). ' and \ need to be escaped.

This fixes the bugs TG-561 and TG-1183.

Corresponding PR: https://github.com/diffblue/test-gen/pull/1195

antlechner committed:
Use UTF-16 conversion function in expr2java

newline and carriage return need to be printed as \n and \r in Java
code, respectively (as opposed to \u000a and \u000d). ' and \ need to be
escaped. The previous implementation was missing those cases. In
addition to that, a call to an existing function is a cleaner way of
writing this.

smowton commented:
I note expr2c (https://github.com/diffblue/cbmc/blob/develop/src/ansi-c/expr2c.cpp#L1791) implements very similar logic, and also escapes `\t`, which we probably ought to here. Suggest factoring that "print ascii char" logic and using it in both "expr2" classes.

antlechner commented:
@smowton I didn't know about `expr2c`. Refactoring this makes sense, where do you suggest I put that function, maybe `cbmc/src/util/unicode.cpp`?
I found with the regression tests I made (see https://github.com/diffblue/test-gen/pull/1195) that escaping `\t` was not necessary in `expr2java` as it converts it to `\u0009` (it also previously converted `\n` to `\u000a` and `\r` to `\u000d`, but those caused problems in Java code, while `\u0009` does not). Still there is probably no harm in escaping `\t`, `\b` and `\f` instead of using their Unicode representation, so I could do that for consistency.

romainbrenguier commented:
I guess `utf16_little_endian_to_java` already does what you need here: https://github.com/diffblue/cbmc/blob/develop/src/util/unicode.cpp#L289

smowton commented:
Ha, forgot about that. Okay, use Romain's suggestion instead of mine.

NathanJPhillips commented:
I would still refactor it into a char version and use that from the string version.

antlechner committed:
Split UTF-16 conversion code into two cases

Previously there was only one case, for strings as an argument. This
commit adds a new function with the same logic applied to a wchar_t
argument. The common code of the two functions is moved to a static
function. Also added an escape for single-quote characters.

antlechner commented:
@romainbrenguier @smowton @NathanJPhillips 
I refactored the existing function in `util/unicode.cpp`. Most of the previous code is now in a static function that operates on `wchar_t`s, and this function is called by both the `wchar_t` and `std::string` versions.
`expr2java` then simply calls the new public `wchar_t` version.
I split the changes into two commits for clarity.
